### ISSUE TITLE
Mac M1 Tensorflow Set Up

### ISSUE THREAD

**Issue by mgebhard** [google/mediapipe#1514]
"ERROR: While resolving toolchains for target @org_tensorflow//tensorflow/tools/git:gen_git_source: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type. Maybe --incompatible_use_cc_configure_from_rules_cc has been flipped and there is no default C++ toolchain added in the WORKSPACE file? See https://github.com/bazelbuild/bazel/issues/10134 for details and migration instructions.
ERROR: Analysis of target '//mediapipe/examples/ios/facedetectioncpu:facedetectioncpu' failed; build aborted: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type. Maybe --incompatible_use_cc_configure_from_rules_cc has been flipped and there is no default C++ toolchain added in the WORKSPACE file? See https://github.com/bazelbuild/bazel/issues/10134 for details and migration instructions."

I am trying to follow instructions here to set up tensorflow using the new mac m1 chip: https://github.com/apple/tensorflow_macos

However, I am still having no luck with initial set up for media pipe.

Have I written custom code: No
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS 11.0.1
Bazel version:nbazel 3.7.2-homebrew
Xcode & Tulsi version (if issue is related to building in mobile dev enviroment): 12.3 Tusli 0.20210106.88
Exact steps to reproduce: ./build_ios_examples.sh -d out_dir --nostrip

**Comment by jiuqiant** [comment]
MediaPipe with its TensorFlow dependence needs Bazel to build everything from scratch. I don't think bazel already supports M1 yet: https://github.com/bazelbuild/bazel/issues/11628.
.

**Comment by mgebhard** [comment]
It looks like the tensorflow team has added M1 support. https://github.com/tensorflow/tensorflow/issues/46178

Is there a workaround to not have to wait for bazel compatibility?

**Comment by jiuqiant** [comment]
I guess not. Also

**Comment by jiuqiant** [comment]
I took a look at https://github.com/tensorflow/tensorflow/commit/e687cab61615a95b8aea79120c9f168d4cc30955 and feel like the mediapipe repo needs some config changes for supporting Apple Silicon. For example, we need to add a config for "macos_arm64"  config_setting like [the one](https://github.com/google/mediapipe/blob/cccf6244d3fc4afc53082a397e764a8560eec96e/mediapipe/BUILD#L113) we have for macos_x86_64. Also, we need to [point to](https://github.com/google/mediapipe/blob/master/WORKSPACE#L366) a TF version that supports Apple Silicon. I guess it won't work as-is.

**Comment by jiuqiant** [comment]
I took a look at https://github.com/tensorflow/tensorflow/commit/e687cab61615a95b8aea79120c9f168d4cc30955. My conclusion is the mediapipe repo needs some config changes for supporting Apple Silicon. For example, the repo needs to add a config for "macos_arm64"  config_setting like [the one](https://github.com/google/mediapipe/blob/cccf6244d3fc4afc53082a397e764a8560eec96e/mediapipe/BUILD#L113) it has for macos_x86_64. Also, it needs to [point to](https://github.com/google/mediapipe/blob/master/WORKSPACE#L366) the latest TF version that supports Apple Silicon.

**Comment by mgebhard** [comment]
Do you have an ETA when the repo will be updated with the config changes? This blocks any integration from the m1 chip

**Comment by mgebhard** [comment]
Any update on M1 support?

**Comment by miholeus** [comment]
also have an error on m1

**Comment by FUBAR2077** [comment]
Yea, not yet I suppose, I cannot install nuget too as it requires x86_64 architecture

**Comment by homuler** [comment]
I confirmed that `hand_tracking_cpu` works on Mac mini (M1, 2020) (macOS Big Sur 11.6) with the following patch applied.
```diff
diff --git a/.bazelrc b/.bazelrc
index 37a0bc1..7d7e159 100644
--- a/.bazelrc
+++ b/.bazelrc
@@ -87,6 +87,11 @@ build:darwin_x86_64 --apple_platform_type=macos
 build:darwin_x86_64 --macos_minimum_os=10.12
 build:darwin_x86_64 --cpu=darwin_x86_64
 
+# May not be necessary
+build:darwin_arm64 --apple_platform_type=macos
+build:darwin_arm64 --macos_minimum_os=10.16
+build:darwin_arm64 --cpu=darwin_arm64
+
 # This bazelrc file is meant to be written by a setup script.
 try-import %workspace%/.configure.bazelrc
 
diff --git a/.bazelversion b/.bazelversion
index 0b2eb36..fae6e3d 100644
--- a/.bazelversion
+++ b/.bazelversion
@@ -1 +1 @@
-3.7.2
+4.2.1
diff --git a/WORKSPACE b/WORKSPACE
index c2aaca6..30afd92 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -200,7 +200,7 @@ new_local_repository(
 new_local_repository(
     name = "macos_opencv",
     build_file = "@//third_party:opencv_macos.BUILD",
-    path = "/usr/local/opt/opencv@3",
+    path = "/opt/homebrew/opt/opencv@3",
 )
 
 new_local_repository(
diff --git a/mediapipe/BUILD b/mediapipe/BUILD
index 1171ea6..72e4208 100644
--- a/mediapipe/BUILD
+++ b/mediapipe/BUILD
@@ -74,6 +74,7 @@ alias(
     name = "macos",
     actual = select({
         ":macos_i386": ":macos_i386",
+        ":macos_arm64": ":macos_arm64",
         ":macos_x86_64": ":macos_x86_64",
         "//conditions:default": ":macos_i386",  # Arbitrarily chosen from above.
     }),
@@ -110,6 +111,15 @@ config_setting(
     visibility = ["//visibility:public"],
 )
 
+config_setting(
+    name = "macos_arm64",
+    values = {
+        "apple_platform_type": "macos",
+        "cpu": "darwin_arm64",
+    },
+    visibility = ["//visibility:public"],
+)
+
 config_setting(
     name = "macos_x86_64",
     values = {
```

The exact command can be found [here](https://google.github.io/mediapipe/getting_started/cpp.html).

It seems that a similar patch is already submitted as a PR.
https://github.com/google/mediapipe/pull/2591

**Comment by Snipeye** [comment]
@homuler Thank you for the recommendations; I've made those changes and I'm still having issues.  Additionally, the PR you linked has issues: It still references bazel 3.7.2.  Additionally in your changes, I had to also change the reference to ffmpeg in WORKSPACE.

I installed the latest bazel (4.2.2) through brew, then updated .bazelversion to reflect that, but when I tried to run the helloworld, I ended up with:

"dyld[70906]: symbol not found in flat namespace '_CFRelease'"

Any further advice?

**Comment by mgebhard** [comment]
It has been almost a year does anyone have a solution to this?

I followed the changes in https://github.com/google/mediapipe/pull/2591/files with slight differences in OpenCV and FFMPEG versions and still got build errors.

```
In file included from mediapipe/framework/formats/image_frame_opencv.cc:15:
In file included from ./mediapipe/framework/formats/image_frame_opencv.h:20:
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
#include <opencv2/core/version.hpp>
```

I would prefer not to use a unity plugin.

**Comment by daviddavid** [comment]
This doesn't solve the issue of running mediapipe on M1 metal, but you can get mediapipe to run on your M1 hardware by using rosetta, as described in [this comment](https://github.com/google/mediapipe/issues/2120#issuecomment-894796864). (Read the rest of the comments before diving in. To solve the homebrew permissions problem in `/usr/local` you can refer to [this SO answer](https://stackoverflow.com/a/46844441))

**Comment by binpan0423** [comment]
> @homuler Thank you for the recommendations; I've made those changes and I'm still having issues. Additionally, the PR you linked has issues: It still references bazel 3.7.2. Additionally in your changes, I had to also change the reference to ffmpeg in WORKSPACE.
> 
> I installed the latest bazel (4.2.2) through brew, then updated .bazelversion to reflect that, but when I tried to run the helloworld, I ended up with:
> 
> "dyld[70906]: symbol not found in flat namespace '_CFRelease'"
> 
> Any further advice?

same problem here。 any solutions？

**Comment by matanster** [comment]
I wonder why they don't reflect any of this limitation on [the installation page at the TensorFlow website](https://www.tensorflow.org/install). They just say this:

> macOS 10.12.6 (Sierra) or later (no GPU support)

Which seems to contradict the actual status if I'm following.

**Comment by justin-f-perez** [comment]
related: #3277 (community M1 wheels)

**Comment by sureshdagooglecom** [comment]
Hi @mgebhard ,
Apple M1 supports in latest release , could you check, verify  and close this issue.

**Comment by anandijain** [comment]
`pip install mediapipe` still gives 
```
ERROR: Could not find a version that satisfies the requirement mediapipe (from versions: none)
ERROR: No matching distribution found for mediapipe
```
I am on Python 3.9.13 and pip 22.1.2.

**Comment by sureshdagooglecom** [comment]
Hi @mgebhard ,
what is the error you are facing w.r.t above issue and please share the error logs.

**Comment by schmidt-sebastian** [comment]
> @sureshdagooglecom `pip install mediapipe` still gives
> 
> ```
> ERROR: Could not find a version that satisfies the requirement mediapipe (from versions: none)
> ERROR: No matching distribution found for mediapipe
> ```
> 
> I am on Python 3.9.13 and pip 22.1.2.

Unfortunately, we still do not have native Python wheels. You can run Python under Rosetta.

**Comment by anandijain** [comment]
thanks! any timeline or eta for when native wheels will be built?

**Comment by schmidt-sebastian** [comment]
Unfortunately, we don't have an ETA for this at this point.

**Comment by drush** [comment]
@schmidt-sebastian Can you enumerate the tasks/gaps that you are aware of for this support to be completed?

**Comment by schmidt-sebastian** [comment]
I don't think I can go into much detail here, but it is an infrastructure issue on our end.

**Comment by The-Real-Thisas** [comment]
This patch seems to work https://github.com/cansik/mediapipe-silicon . Would it be possible to implement this so there can be a wheel for people to download without having to run through rosetta. 

In the meanwhile for people wanting to install mediapipe I suggest installing one of the prebuilt binaries with `pip install` https://github.com/cansik/mediapipe-silicon/releases/tag/v0.8.10.1

For me `pip install mediapipe_silicon-0.8.10.1-cp39-cp39-macosx_12_0_arm64.whl` worked.

**Comment by The-Real-Thisas** [comment]
Hi @sureshdagooglecom just so you know what I mentioned is just a workaround and is not a pull request or anything to fix it mainline. I don't think this issue is resolved.

**Comment by davidcavazos** [comment]
I was able to install `tensorflow-macos` (not `tensorflow`) following Apple's instructions here: https://developer.apple.com/metal/tensorflow-plugin/

However, when I try to `pip install` other packages requiring `grpcio` like `google-cloud-aiplatform`, I get errors while trying to compile `grpcio`. This has been rather frustrating, and I would expect TensorFlow working with Vertex AI running from my M1 MacBook Pro.

**Comment by kuaashish** [comment]
Hello @mgebhard,
We have released Universal Wheels with version 0.9.3.0 as guided [here](https://github.com/google/mediapipe/releases/tag/v0.9.3.0) in the release page with Python 3.11.
Please try out to build mediapipe using pip let us know if still an issue from your end. Thank you
